### PR TITLE
WIP: Contact improvements

### PR DESCRIPTION
**Pull request by kain88-de**
This continues the work in #708.  I have stripped down `Contacts` to the absolute bare essential (besides `single_frame` everything is gone, renamed or moved to it's own function). I think this results in a cleaner implementation in the end and also one that is more flexible. I'm aware that I might have removed useful functionality but I'm happy to add that again if someone explains to me why they need it.

I've created a ipython notebook in the root-directory where I outlined the normal fraction of native contacts analysis and a new analysis for an easy binary decision algorithm if a protein is folded or two partners are bound. I plan to keep this notebook always at the top comment and remove it before I merge. I would like some help if this is all the usage that people want and what use cases are currently missing as example, or if they can't be done with the class right now)

The plots are now removed as well because I don't think the `plot_qavg` contained a lot of content (basically only a complicated way to visualize `np.mean(q)`). The timeseries plot is removed because with arbitrary functions allowed as methods it might now represent accurately the data that is plotted. 

Changes made in this Pull Request:
 -  removed all untested code from `Contacts`
 -  Contacts now accepts arbitrary functions as method kwarg
 -  old contact analysis classes are moved to be deprecated soon

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - ~~[ ] Issue raised/referenced?~~
 - [x] deprecate old contact classes
 - ~~[ ] pbc keyword~~

### CODE REVIEW

File: Notebooks/test new class.ipynb [hunk 1]
```diff
@@ -0,0 +1,276 @@
+{
```
Comment (richardjgowers) [inline Notebooks/test new class.ipynb]: I think we moved all the examples out into the cookbook, so this should probably go there.

File: package/MDAnalysis/analysis/contacts.py [hunk 2]
```diff
@@ -164,6 +176,199 @@
...
+        the same size as distances
+
+    Returns
+    -------
+    Q : float
+        fraction of distances below cutoff
+
+    """
+    y = distances <= cutoff
+    return float(y.sum()) / distances.size
```
Comment (richardjgowers) [inline package/MDAnalysis/analysis/contacts.py]: Would it be better to start using `__future__` division throughout the module?

File: package/MDAnalysis/analysis/contacts.py [hunk 3]
```diff
@@ -164,6 +176,199 @@
...
+        ----------
+        results: list
+            Fraction of native contacts for each frame
+        """
+        if method == 'cutoff':
+            self.fraction_contacts = hard_cut
+        elif method == 'best-hummer':
+            self.fraction_contacts = best_hummer_q
+        else:
+            self.fraction_contacts = method
```
Comment (richardjgowers) [inline package/MDAnalysis/analysis/contacts.py]: Is this letting you put in a custom function?  If so document it

File: package/MDAnalysis/analysis/contacts.py [hunk 4]
```diff
@@ -164,6 +176,199 @@
...
+        refA, refB = refgroup
+        self.r0 = distance_array(refA.positions, refB.positions)
+        self.initial_contacts = contact_matrix(self.r0, radius)
+
+        self.fraction_kwargs = kwargs
+        self.timeseries = []
+
+    def _single_frame(self):
+        # compute distance array for a frame
+        d = distance_array(self.grA.positions, self.grB.positions)
```
Comment (richardjgowers) [inline package/MDAnalysis/analysis/contacts.py]: pbc?

File: package/MDAnalysis/analysis/contacts.py [hunk 5]
```diff
@@ -164,6 +176,199 @@
...
+
+        # select only the contacts that were formed in the reference state
+        r = d[self.initial_contacts]
```
Comment (richardjgowers) [inline package/MDAnalysis/analysis/contacts.py]: Could we make the distance_array call above lighter by applying the mask there instead of here? (`grA[self.initial_contacts].positions`)

Comment (kain88-de) [inline package/MDAnalysis/analysis/contacts.py]: Good point. Haven't looked into that yet.

File: package/MDAnalysis/analysis/contacts.py (hunk 4 above)
Comment (kain88-de) [inline package/MDAnalysis/analysis/contacts.py]: yup another good point.

File: package/MDAnalysis/analysis/contacts.py (hunk 3 above)
Comment (kain88-de) [inline package/MDAnalysis/analysis/contacts.py]: Yes. I haven't documented much of it yet because I want to wait until others tell me that my usecase examples are good or crap and I have to change the API again.

File: package/MDAnalysis/analysis/contacts.py (hunk 2 above)
Comment (kain88-de) [inline package/MDAnalysis/analysis/contacts.py]: Very good idea.

File: Notebooks/test new class.ipynb (hunk 1 above)
Comment (kain88-de) [inline Notebooks/test new class.ipynb]: Yeah but I think this way it is currently easier to have the use-case notebook with the changes in the contacts class. This is thought of a convenience for me and testers. I will remove this before I merge the PR and collect the examples in the module docstring.

**Comment by kain88-de** [comment]
I had a look for the suggested speed improvements in `_single_frame` and to add `pbc` for the distance calculations. While both are possible I don't think I have time to implement both right now, this is because the *minimal image* functions for the different pbc boxes aren't accessible in python-code right now.

I would use the slow `distance_array` for now and add a `pbc` keyword. The speed improvements could be done later.

**Comment by richardjgowers** [comment]
@kain88-de no worries.  Which function needs access to minimum image?

**Comment by kain88-de** [comment]
[minimum_image_triclinic](https://github.com/MDAnalysis/mdanalysis/blob/bddbd81c6e784aceb8d38b51b211ba1dce4c66bd/package/MDAnalysis/lib/include/calc_distances.h#L45) and [minimum_image](https://github.com/MDAnalysis/mdanalysis/blob/bddbd81c6e784aceb8d38b51b211ba1dce4c66bd/package/MDAnalysis/lib/include/calc_distances.h#L33), plus a nice python wrapper that selects the appropriate one of the two depending on the box.

**Comment by richardjgowers** [comment]
I can't see where you need to use those? Aren't you just using
distance_array for all the distances in this module?

**Comment by kain88-de** [comment]
Not when I want to implement your speed improvement suggestions

```python
def  _single_frame(self):
    r = np.linalg.norm(self.grpA.positions[self.idx_grpA] - self.grpB.positions[self.idx_grpB], axis=1)
    r0 = self.r0[self.initial_contacs]
    ...
```

Where `self.idx_grp*` is calculated in `__init__` based on the initial contacts. For the test case example this reduces the distance calculations for each from from 40*70 to 60, an order of magnitude less then pure `distance_array`!

But to have this also with periodic boundary conditions I need a function to calculate the minimal image convention distance before I calculate the norm.

**Comment by richardjgowers** [comment]
So you want to zip over the distances? Calc_bonds does this

**Comment by richardjgowers** [comment]
Anyway is this ready to merge?

**Comment by kain88-de** [comment]
> Anyway is this ready to merge?

No docs are missing. There is still a bit of work left. I have to do some other stuff for my PhD right now. I hope I have some time next week to get some things done. The big 4 that still need doing are [docs, tests, pbc, deprecations for old classes].

While we are at it. What is a good deprecation release?

@jandom  Does this still work for you now? I removed quite a lot of things.

@orbeckst You were against removing the old contact classes because of some missing functionality. Could you tell me what that was and how it was used, the old code is not that easy to understand.

**Comment by orbeckst** [comment]
1. Deprecated `ContactAnalysis1` should go eventually (in 0.15 or later?), @jandom 's `Contacts` is the new hotness.
2. `ContactAnalysis` does a *q1-q2* analysis with native contacts relative to an initial and a final state. This functionality is not available in `Contacts` (unless you find a way to compute native contacts for multiple reference structures). It's fugly, without a question, but it does [get used](http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004568)...

File: package/MDAnalysis/analysis/contacts.py (hunk 3 above)
Comment (orbeckst) [inline package/MDAnalysis/analysis/contacts.py]: If you allow a custom function then I would check at least that it is [callable](https://docs.python.org/2/library/functions.html#callable).

You can also collect all the existing functions in a dispatch table and just do
```python
try:
  self.fraction_contacts = METHODS[method]
except KeyError:
  if callable(method):
    self.fraction_contacts = method
  else:
    raise TypeError("method must be ...")
```

**Comment by kain88-de** [comment]
> unless you find a way to compute native contacts for multiple reference structures

This means I have to allow several `(ref_groupA, ref_groupB)` tuples to be passed to the initialization?

If I understand you correctly the `q1-q2` analysis would currently be calculated going over the trajectory twice.

```python
refA = u.select_atoms('...')
refB = u.setlect_atoms('...')

q1 = contacts.Contacts(u, (selA, selB), (refA, refB))

u.trajectory[-1]
q2 = contacts.Contacts(u,  (selA, selB), (refA, refB))
```

What you would want is something like this?

```python
q = contacts.Contacts(u, (selA, selB), ((refA_initial, refB_initial),
                                        (refA_final, refB_final)))
```

If that is correct it should be easy to extend. The preparation of the reference AtomGroup is a little bit more involved though then the current one. Essentially you would need to load the Universe twice and go to the last frame once. This would be made easier if we could get `freeze_copy` of a AtomGroup so that it unlinks from any trajectory/universe object.

**Comment by orbeckst** [comment]
Yes to all. 

--
Oliver Beckstein
email: orbeckst@gmail.com


Am Mar 26, 2016 um 8:57 schrieb kain88-de <notifications@github.com>:

> unless you find a way to compute native contacts for multiple reference structures
> 
> This means I have to allow several (ref_groupA, ref_groupB) tuples to be passed to the initialization?

**Comment by kain88-de** [comment]
> Yes to all.

Nice. That shouldn't be complicated.

I'm only waiting for @jandom to answer if I deleted anything he really needed.

File: package/MDAnalysis/analysis/contacts.py (hunk 3 above)
Comment (kain88-de) [inline package/MDAnalysis/analysis/contacts.py]: Yes I should to that.

**Comment by kain88-de** [comment]
It is now possible to create several reference groups to compare against (not only two). And I have added a small wrapper to quickly create a q1-q2 analysis.

@orbeckst Have a look at the [notebook](https://github.com/MDAnalysis/mdanalysis/blob/contact-improvements/Notebooks/test%20new%20class.ipynb). The q1q2 analysis is in the last section. Does this satisfy you?

**Comment by orbeckst** [comment]
The q1-q2 looks good. 

As a last test, could you please also run it as the original example http://www.mdanalysis.org/mdanalysis/documentation_pages/analysis/contacts.html#two-dimensional-contact-analysis-q1-q2 (you might need to reverse-engineer the default values for the selections, I can't remember if these were C-alphas or all heavy or all backbone) and see if it reproduces the figure http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3144279/figure/F5/ ?

**Comment by kain88-de** [comment]
yeah that doesn't work at all -.-. I've updated the notebook and I'm quite sure about the selections since the old code matches with the publication figure.

But When I look at the code the old contact analysis does things a little bit different from me. I'm not sure why that is. I appreciate some help with this.

**Comment by orbeckst** [comment]
Did you post the output from your code for the AdK problem? I did not see it (only the salt bridge example).

**Comment by kain88-de** [comment]
I've updated the notebook, changes are at the bottom. It only uses the *CA* atoms now and I'm comparing to the vanilla output of ContactAnalysis (which looks similar to the figure from the paper).

**Comment by orbeckst** [comment]
Are you using the same cutoff?

**Comment by kain88-de** [comment]
Yes 8 Angström in both cases. I suspect that the calculation is slightly different.

**Comment by kain88-de** [comment]
@orbeckst I should have fixed the q1q2 wrapper function now. The problem was that the hard_cut_q doesn't calculate the same as the old `ContactAnalysis`. In short the old class counted a native contact it if was formed in the reference and the distance shorter then some radius *r*. The normal **hard_cut_q** will only count a native contact if the distance is closer then in the reference. I updated the notebook as weil for you to look at.

This brings me to another thing I would like some help with naming.

- **hard_cut_q** count a contact if the distance is closer then the reference distance
- **min_cut** count a contact if the distance is smaller then some radius *r* in the current frame and closer then *r* in the reference
- **soft_cut_q** do a soft-potential cut on the native contacts formed in the reference, being 0.5 at the exact refreence distance

I think **min_cut** is wrong but I can't think of anything better.

**Comment by orbeckst** [comment]
The updated q<sub>1</sub>-q<sub>2</sub> looks good now. 
What you named **min_cut** is what I had found as **fraction of native contacts** in 

J. Franklin, P. Koehl, S. Doniach, and M. Delarue. MinActionPath: maximum likelihood trajectory for large- scale structural transitions in a coarse-grained locally harmonic energy landscape. Nucleic Acids Res., 35 (Web Server issue):W477–W482, 2007. doi: [10.1093/nar/gkm342](http://doi.org/10.1093/nar/gkm342).

Maybe just call it `FKDM_q` or `Franklin_Koehl_Doniach_Delarue_q`?

**soft_cut_q** is still Best-Hummer, right? Maybe keep the explicit name because there are, in principle, many ways to implement a soft cutoff. (mdtraj called it [best_hummer_q](http://mdtraj.org/latest/examples/native-contact.html)).

**Comment by kain88-de** [comment]
I don't like to give this stuff the author names. It requires people to know the literature very well. I can imagine some new students that don't use this because they don't read the docs and don't know the names or papers, so they don't have any idea of what the function does and so rather go with what they know from the name). I rather have some more intuitive name that gives a sense of what is being done.

**Comment by orbeckst** [comment]
On 4 Apr, 2016, at 23:34, kain88-de wrote:

> I don't like to give this stuff the author names. It requires people to know the literature very well. I can imagine some new students that don't use this because they don't read the docs and don't know the names or papers, so they don't have any idea of what the function does and so rather go with what they know from the name). I rather have some more intuitive name that gives a sense of what is being done.

Fair enough.

**Comment by kain88-de** [comment]
I've gotten a little bit further. Here is an overview of what is still to do. I'm happy for help as this really has to go into 0.15.0 because the current implementation doesn't work correct, all of which is fixed in this version.

Questions
--------------
1. What do we need to cite? I have the Best & Hummer paper and the Franklin paper so far. Am I missing someone?
2. Should we only calculate the fraction of contacts or also the absolute number of contacts?

Still TODO
---------------
- update module documentation (notebook should serve as example reference)
- write new tests for updated class
 anything else?

Saved for Later
---------------------
- Periodic Boundary Conditions (lets do that in a separate issue)
- performance improvements

**Comment by orbeckst** [comment]
Re questions: 
1. That's the only two references that I remember looking at.
2. Typically, fractions are reported, aren't they? I think if we need absolute numbers we'll raise an issue...

Main thing is docs and fixing tests (are you getting rid of ContactAnalysis1 because then the test failures will go away...). I would drop ContactAnalysis1 cold turkey and only leave a stub in place that says that this functionality is now provided by .... and link to the docs.

**Comment by kain88-de** [comment]
So docs are done. I have included examples for the normal usage , the q1q2 analysis and the new option to pass custom method functions. Comments are welcome.

File: package/MDAnalysis/analysis/contacts.py [hunk 6]
```diff
@@ -14,45 +14,35 @@
...
-analysis:
+1. *Hard Cut*: To count as a contact the atoms *i* and *j* have to be at least
+   as close as in the reference structure.
 
-1. Contacts between two groups of atoms are defined with
-   :class:`ContactAnalysis1`), which allows one to calculate *q(t)* over
-   time. This is especially useful in order to look at native contacts during
-   an equilibrium simulation where one can also look at the average matrix of
-   native contacts (see :meth:`ContactAnalysis1.plot_qavg`).
+2. *Soft Cut*: Tha atom pair *i* and *j* is assigned based on a soft potential
```
Comment (richardjgowers) [inline package/MDAnalysis/analysis/contacts.py]: The

File: package/MDAnalysis/analysis/contacts.py [hunk 7]
```diff
@@ -14,45 +14,35 @@
...
 
-The "fraction of native contacts" *q(t)* is a number between 0 and 1 and
-calculated as the total number of native contacts for a given time frame
-divided by the total number of contacts in the reference structure.
+Analysis of native contacts *Q* over a trajectory. Native contacts of a
+conformation are contacts that exist in a reference structure and in the
+conformation. Contacts in the reference structure are always defined as being
+closer then a distance `radius`. The fraction of native contacts for a
+conformation can be calculated in different ways. This module supports 3
+different ones
```
Comment (richardjgowers) [inline package/MDAnalysis/analysis/contacts.py]: maybe something like "This module supports 3 different metrics listed below, as well as custom metrics", then you can remove lines 43-45 as that information is now here

File: package/MDAnalysis/analysis/contacts.py [hunk 8]
```diff
@@ -64,86 +54,133 @@
...
+>>> import MDAnalysis as mda
+>>> from MDAnalysis.analysis import contacts
+>>> from MDAnalysis.tests.datafiles import PSF,DCD
+>>> import matplotlib.pyplot as plt
+>>> # example trajectory (transition of AdK from closed to open)
+>>> u = mda.Universe(PSF,DCD)
+>>> # crude definition of salt bridges as contacts between NH/NZ in ARG/LYS and
+>>> # OE*/OD* in ASP/GLU. You might want to think a little bit harder about the
+>>> # problem before using this for real work.
+>>> sel_basic = "(resname ARG or resname LYS) and (name NH* or name NZ)"
```
Comment (richardjgowers) [inline package/MDAnalysis/analysis/contacts.py]: These could use the new implicit or, so `"(resname ARG LYS) and (name NH* NZ)"`

File: package/MDAnalysis/analysis/contacts.py [hunk 9]
```diff
@@ -152,22 +189,310 @@
...
+def soft_cut_q(r, r0, beta=5.0, lambda_constant=1.8):
+    r"""Calculate fraction of native contacts *Q* for a soft cut off
+
+    ..math::
+        Q(r, r_0) = \frac{1}{1 + e^{\beta (r - \lambda r_0)}}
+
+    Reasonable values for different simulation types are
+
+    *All Atom*: lambda_constant = 1.8 (unitless)
+    *Coarse Grained*: lambda_constant = 1.8 (unitless)
```
Comment (richardjgowers) [inline package/MDAnalysis/analysis/contacts.py]: Are these meant to be the same?

**Comment by richardjgowers** [comment]
Looks good to me apart from a few doc comments, just remember to make a new issue for PBC for this as it's a nice starter issue for someone.

File: package/MDAnalysis/analysis/contacts.py (hunk 9 above)
Comment (kain88-de) [inline package/MDAnalysis/analysis/contacts.py]: Nope they aren't thank.

**Comment by kain88-de** [comment]
OK on my side this is done now. The docs are updated for `Contacts` and there are more tests, the examples have been added as tests as well. For the changelog I have updated the message from @jandom.

@jandom any comments on the changes I made?

**Comment by orbeckst** [comment]
This seems good to go. One last ping @jandom but if there are not other comments over the next day we can merge.

@kain88-de did you raise an issue for the PBC specific stuff and referenced this PR (just to have a trail of breadcrumbs)?

**Comment by kain88-de** [comment]
Not yet I will do once this is merged.
